### DISCUSSION TITLE
Objective function

### DISCUSSION THREAD

**Discussion by Charly-28** [mdolab/dafoam#131] [General]
Hello, 
I have a few questions regarding Dafoam's capability.

1) Is Dafoam able to do multi-objective obtimization? For example the thrust (force on a propeller) and the torque.
2) Could we do an aeroacoustic optimization? I think that for this question it will be necessary to create an objective function.
3) How integrated a new objective function in DaFoam?
Thanks for your help.

Charly

**Comment by friedenhe** [comment]
1. Yes, it is very straightforward to set up multi-objective optimization in DAFoam, check [this](https://github.com/DAFoam/tutorials/blob/9829d6bd851c5eef779c897d8579611a27fe1311/UBend_Channel/runScript.py#L41), it is a multi-objective setup with total pressure loss and heat transfer. You can follow it to define an objective function that has multiple parts, part 1 could be the thrust, and part 2 could be the torque. Remember to give them proper weights.
2. No, we don't have an acoustic model in DAFoam yet, but you are welcome to add such an objective function.
3. To add a new objective function, you need to add a new child class in `src/adjoint/DAObjFunc` to implement how to compute this objective function. Refer to the implementations for other objective functions.

**Reply by hosseinnegahban** [reply]
> 1. Yes, it is very straightforward to set up multi-objective optimization in DAFoam, check [this](https://github.com/DAFoam/tutorials/blob/9829d6bd851c5eef779c897d8579611a27fe1311/UBend_Channel/runScript.py#L41), it is a multi-objective setup with total pressure loss and heat transfer. You can follow it to define an objective function that has multiple parts, part 1 could be the thrust, and part 2 could be the torque. Remember to give them proper weights.
> 2. No, we don't have an acoustic model in DAFoam yet, but you are welcome to add such an objective function.
> 3. To add a new objective function, you need to add a new child class in `src/adjoint/DAObjFunc` to implement how to compute this objective function. Refer to the implementations for other objective functions.

Hi,
Could we add this new class in DAFoam image or we have to install it from source? 
I want to add the lift to drag ratio objective function. I would appreciate it if you help me in this regard.

Thank you
Hossein

**Reply by friedenhe** [reply]
You have to implement it by youself! We will have a workshop this summer to talk about DAFoam code development. Feel free to join if you are interested.

**Comment by Charly-28** [comment]
Thanks for the answers. They will help me especially for the implementation of objective aeroacoustic functions. I have another question concerning the instalation.
1) Is Dafoam compatible with the recent version of OpenFoam (Openfoam v2012).
2) Could it be installed on Ubuntu 20.04 or the most recent version?
Thanks

Charly

**Reply by friedenhe** [reply]
Nope... we don't support OpenFOAM-v2012. And yes, you can compile dafoam on Ubuntu 20.04. We just updated the [installation guide](https://dafoam.github.io/mydoc_installation_source.html) for Ubuntu 20.04. There is only one minor change when installing Miniconda.

**Comment by Charly-28** [comment]
Thanks for your reply. 
But I have a problem installing numpy 1.14.3. Do you have an idea?
Below are the installation errors: 

pip install numpy==1.14.3
Collecting numpy==1.14.3
  Using cached numpy-1.14.3.zip (4.9 MB)
Building wheels for collected packages: numpy
  Building wheel for numpy (setup.py) ... error
  ERROR: Command errored out with exit status 1:
   command: /home/kone/dafoam/packages/miniconda3/bin/python -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-_l2h8yrn/numpy_ef83e22b66a44ffcad649a884a9864d8/setup.py'"'"'; __file__='"'"'/tmp/pip-install-_l2h8yrn/numpy_ef83e22b66a44ffcad649a884a9864d8/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-rgge2how
       cwd: /tmp/pip-install-_l2h8yrn/numpy_ef83e22b66a44ffcad649a884a9864d8/
  Complete output (307 lines):
  Running from numpy source directory.

Thanks for your help.
Charly

**Reply by friedenhe** [reply]
Try: `pip install numpy==1.19.5`

**Reply by Charly-28** [reply]
Thanks for your reply. Now numpy is run. But I have a problem installing idward 2.2.1. 
Below are installation errors:

In file included from /home/kone/dafoam/packages/miniconda3/lib/python3.7/site-packages/numpy/core/include/numpy/ndarraytypes.h:1822,
                 from /home/kone/dafoam/packages/miniconda3/lib/python3.7/site-packages/numpy/core/include/numpy/ndarrayobject.h:12,
                 from /home/kone/dafoam/packages/miniconda3/lib/python3.7/site-packages/numpy/core/include/numpy/arrayobject.h:4,
                 from /home/kone/dafoam/packages/miniconda3/lib/python3.7/site-packages/numpy/f2py/src/fortranobject.h:13,
                 from idwarpmodule.c:16:
/home/kone/dafoam/packages/miniconda3/lib/python3.7/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:17:2: warning: #warning "Using deprecated NumPy API, disable it with " "#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
   17 | #warning "Using deprecated NumPy API, disable it with " \
      |  ^~~~~~~
idwarpmodule.c: In function ‘f2py_setup_cgnsgrid’:
idwarpmodule.c:3231:36: warning: assignment to ‘f2py_init_func’ {aka ‘void (*)(int *, long int *, void (*)(char *, long int *), int *)’} from incompatible pointer type ‘void (*)(int *, int *, void (*)(char *, int *), int *)’ [-Wincompatible-pointer-types]
 3231 |   f2py_cgnsgrid_def[i_f2py++].func = surfacepoints;
      |                                    ^
idwarpmodule.c:3232:36: warning: assignment to ‘f2py_init_func’ {aka ‘void (*)(int *, long int *, void (*)(char *, long int *), int *)’} from incompatible pointer type ‘void (*)(int *, int *, void (*)(char *, int *), int *)’ [-Wincompatible-pointer-types]
 3232 |   f2py_cgnsgrid_def[i_f2py++].func = surfaceconn;
      |                                    ^
idwarpmodule.c:3233:36: warning: assignment to ‘f2py_init_func’ {aka ‘void (*)(int *, long int *, void (*)(char *, long int *), int *)’} from incompatible pointer type ‘void (*)(int *, int *, void (*)(char *, int *), int *)’ [-Wincompatible-pointer-types]
 3233 |   f2py_cgnsgrid_def[i_f2py++].func = surfaceiswall;
      |                                    ^
idwarpmodule.c:3234:36: warning: assignment to ‘f2py_init_func’ {aka ‘void (*)(int *, long int *, void (*)(char *, long int *), int *)’} from incompatible pointer type ‘void (*)(int *, int *, void (*)(char *, int *), int *)’ [-Wincompatible-pointer-types]
 3234 |   f2py_cgnsgrid_def[i_f2py++].func = surfaceissymm;
      |                                    ^
idwarpmodule.c:3235:36: warning: assignment to ‘f2py_init_func’ {aka ‘void (*)(int *, long int *, void (*)(char *, long int *), int *)’} from incompatible pointer type ‘void (*)(int *, int *, void (*)(char *, int *), int *)’ [-Wincompatible-pointer-types]
 3235 |   f2py_cgnsgrid_def[i_f2py++].func = surfacesizes;
      |                                    ^
idwarpmodule.c:3236:36: warning: assignment to ‘f2py_init_func’ {aka ‘void (*)(int *, long int *, void (*)(char *, long int *), int *)’} from incompatible pointer type ‘void (*)(int *, int *, void (*)(char *, int *), int *)’ [-Wincompatible-pointer-types]
 3236 |   f2py_cgnsgrid_def[i_f2py++].func = surfaceptr;
      |                                    ^
idwarpmodule.c:3237:36: warning: assignment to ‘f2py_init_func’ {aka ‘void (*)(int *, long int *, void (*)(char *, long int *), int *)’} from incompatible pointer type ‘void (*)(int *, int *, void (*)(char *, int *), int *)’ [-Wincompatible-pointer-types]
 3237 |   f2py_cgnsgrid_def[i_f2py++].func = surfacepatchptr;
      |                                    ^
idwarpmodule.c:3238:36: warning: assignment to ‘f2py_init_func’ {aka ‘void (*)(int *, long int *, void (*)(char *, long int *), int *)’} from incompatible pointer type ‘void (*)(int *, int *, void (*)(char *, int *), int *)’ [-Wincompatible-pointer-types]
 3238 |   f2py_cgnsgrid_def[i_f2py++].func = surfacenames;
      |                                    ^
idwarpmodule.c: In function ‘f2py_setup_plot3dsurface’:
idwarpmodule.c:3257:41: warning: assignment to ‘f2py_init_func’ {aka ‘void (*)(int *, long int *, void (*)(char *, long int *), int *)’} from incompatible pointer type ‘void (*)(int *, int *, void (*)(char *, int *), int *)’ [-Wincompatible-pointer-types]
 3257 |   f2py_plot3dsurface_def[i_f2py++].func = pts;
      |                                         ^
idwarpmodule.c:3258:41: warning: assignment to ‘f2py_init_func’ {aka ‘void (*)(int *, long int *, void (*)(char *, long int *), int *)’} from incompatible pointer type ‘void (*)(int *, int *, void (*)(char *, int *), int *)’ [-Wincompatible-pointer-types]
 3258 |   f2py_plot3dsurface_def[i_f2py++].func = conn;
      |                                         ^
mpicc -I../../mod -I/home/kone/dafoam/packages/CGNS-3.3.0/opt-gfortran/include -fPIC -O2 -I/home/kone/dafoam/packages/petsc-3.11.4/include -I/home/kone/dafoam/packages/petsc-3.11.4/real-opt/include -I/home/kone/dafoam/packages/petsc-3.11.4 -I/home/kone/dafoam/packages/miniconda3/include/python3.7m -I/home/kone/dafoam/packages/miniconda3/include/python3.7m -I/home/kone/dafoam/packages/miniconda3/lib/python3.7/site-packages/numpy/core/include -c \
/home/kone/dafoam/packages/miniconda3/lib/python3.7/site-packages/numpy/f2py/src/fortranobject.c -o fortranobject.o
In file included from /home/kone/dafoam/packages/miniconda3/lib/python3.7/site-packages/numpy/core/include/numpy/ndarraytypes.h:1822,
                 from /home/kone/dafoam/packages/miniconda3/lib/python3.7/site-packages/numpy/core/include/numpy/ndarrayobject.h:12,
                 from /home/kone/dafoam/packages/miniconda3/lib/python3.7/site-packages/numpy/core/include/numpy/arrayobject.h:4,
                 from /home/kone/dafoam/packages/miniconda3/lib/python3.7/site-packages/numpy/f2py/src/fortranobject.h:13,
                 from /home/kone/dafoam/packages/miniconda3/lib/python3.7/site-packages/numpy/f2py/src/fortranobject.c:2:
/home/kone/dafoam/packages/miniconda3/lib/python3.7/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:17:2: warning: #warning "Using deprecated NumPy API, disable it with " "#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
   17 | #warning "Using deprecated NumPy API, disable it with " \
      |  ^~~~~~~
mpif90 -I../../mod -I/home/kone/dafoam/packages/CGNS-3.3.0/opt-gfortran/include -DUSECGNSMODULE   -fPIC -fPIC -fdefault-real-8 -O2 -I/home/kone/dafoam/packages/petsc-3.11.4/include -I/home/kone/dafoam/packages/petsc-3.11.4/real-opt/include -I/home/kone/dafoam/packages/petsc-3.11.4 -I../../mod -c idwarp-f2pywrappers2.f90 
mpif90 -shared fortranobject.o idwarpmodule.o idwarp-f2pywrappers2.o   -L../../lib -lidwarp -Wl,-rpath,/home/kone/dafoam/packages/petsc-3.11.4/real-opt/lib -L/home/kone/dafoam/packages/petsc-3.11.4/real-opt/lib -Wl,-rpath,/home/kone/dafoam/packages/petsc-3.11.4/real-opt/lib -L/home/kone/dafoam/packages/petsc-3.11.4/real-opt/lib -Wl,-rpath,/home/kone/dafoam/OpenFOAM/ThirdParty-v1812/platforms/linux64Gcc/openmpi-1.10.4/lib64 -L/home/kone/dafoam/OpenFOAM/ThirdParty-v1812/platforms/linux64Gcc/openmpi-1.10.4/lib64 -Wl,-rpath,/usr/lib/gcc/x86_64-linux-gnu/9 -L/usr/lib/gcc/x86_64-linux-gnu/9 -Wl,-rpath,/usr/lib/x86_64-linux-gnu -L/usr/lib/x86_64-linux-gnu -Wl,-rpath,/lib/x86_64-linux-gnu -L/lib/x86_64-linux-gnu -lpetsc -lsuperlu_dist -lflapack -lfblas -lparmetis -lmetis -lm -lX11 -lstdc++ -ldl -lmpi_usempif08 -lmpi_usempi_ignore_tkr -lmpi_mpifh -lmpi -lgfortran -lm -lgfortran -lm -lgcc_s -lquadmath -lpthread -lstdc++ -ldl  -L/home/kone/dafoam/packages/CGNS-3.3.0/opt-gfortran/lib -lcgns -o idwarp.so
python importTest.py
Testing if module idwarp can be imported...
Error: idwarp was not imported correctly
make[2]: *** [Makefile:42: all] Error 1
make[2]: Leaving directory '/home/kone/dafoam/repos/idwarp-2.2.1/src/f2py'
make[1]: *** [Makefile:57: warp] Error 2
make[1]: Leaving directory '/home/kone/dafoam/repos/idwarp-2.2.1'
make: *** [Makefile:24: default] Error 2

Thanks
Charly

**Reply by friedenhe** [reply]
Try numpy-1.16

**Reply by Charly-28** [reply]
I still have the same errors: 

python importTest.py
Testing if module idwarp can be imported...
Error: idwarp was not imported correctly
make[2]: *** [Makefile:42: all] Error 1
make[2]: Leaving directory '/home/kone/dafoam/repos/idwarp-2.2.1/src/f2py'
make[1]: *** [Makefile:57: warp] Error 2
make[1]: Leaving directory '/home/kone/dafoam/repos/idwarp-2.2.1'
make: *** [Makefile:24: default] Error 2

**Reply by friedenhe** [reply]
Run `find . -name idwarp.so` to check if idwarp.so is generated. If yes, run `ldd idwarp.so`, and what is the output? Also, source `loadDAFoam.sh`, go to the folder where idwarp.so is located and run `python`, then run `import idwarp` and see if you get some error messages. In addition, check if you have some existing modules in the `.local/lib/python3.7/site-packages` folder because they may conflict with idwarp. Lastly, check if the CGNS lib is properly compiled. You should find libcgns.so in CGNS-3.3.0/opt-fortran/lib

**Reply by Charly-28** [reply]
Below the output of the comments: 

1) find . -name idwarp.so
./src/f2py/idwarp.so

2) ldd idwarp.so
/bin/bash: /home/kone/dafoam/packages/miniconda3/lib/libtinfo.so.6: no version information available (required by /bin/bash)
	linux-vdso.so.1 (0x00007ffdae927000)
	libpetsc.so.3.11 => /home/kone/dafoam/packages/petsc-3.11.4/real-opt/lib/libpetsc.so.3.11 (0x00007f4187810000)
	libgfortran.so.5 => /usr/lib/x86_64-linux-gnu/libgfortran.so.5 (0x00007f4187548000)
	libm.so.6 => /usr/lib/x86_64-linux-gnu/libm.so.6 (0x00007f41873f7000)
	libmpi_mpifh.so.12 => /home/kone/dafoam/OpenFOAM/ThirdParty-v1812/platforms/linux64Gcc/openmpi-1.10.4/lib64/libmpi_mpifh.so.12 (0x00007f4187395000)
	libgcc_s.so.1 => /home/kone/dafoam/packages/miniconda3/lib/libgcc_s.so.1 (0x00007f4187381000)
	libcgns.so.3.4 => /usr/lib/x86_64-linux-gnu/libcgns.so.3.4 (0x00007f41872b3000)
	libmpi_usempif08.so.11 => /home/kone/dafoam/OpenFOAM/ThirdParty-v1812/platforms/linux64Gcc/openmpi-1.10.4/lib64/libmpi_usempif08.so.11 (0x00007f418727b000)
	libc.so.6 => /usr/lib/x86_64-linux-gnu/libc.so.6 (0x00007f4187089000)
	libsuperlu_dist.so.6 => /home/kone/dafoam/packages/petsc-3.11.4/real-opt/lib/libsuperlu_dist.so.6 (0x00007f4186fac000)
	libparmetis.so => /home/kone/dafoam/packages/petsc-3.11.4/real-opt/lib/libparmetis.so (0x00007f4186f69000)
	libmetis.so => /home/kone/dafoam/packages/petsc-3.11.4/real-opt/lib/libmetis.so (0x00007f4186efc000)
	libX11.so.6 => /usr/lib/x86_64-linux-gnu/libX11.so.6 (0x00007f4186dbf000)
	libdl.so.2 => /usr/lib/x86_64-linux-gnu/libdl.so.2 (0x00007f4186db9000)
	libmpi.so.12 => /home/kone/dafoam/OpenFOAM/ThirdParty-v1812/platforms/linux64Gcc/openmpi-1.10.4/lib64/libmpi.so.12 (0x00007f4186cca000)
	libpthread.so.0 => /usr/lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f4186ca7000)
	libquadmath.so.0 => /home/kone/dafoam/packages/miniconda3/lib/libquadmath.so.0 (0x00007f4186c6d000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f4188a99000)
	libopen-pal.so.13 => /home/kone/dafoam/OpenFOAM/ThirdParty-v1812/platforms/linux64Gcc/openmpi-1.10.4/lib64/libopen-pal.so.13 (0x00007f4186b98000)
	libhdf5_serial.so.103 => /lib/x86_64-linux-gnu/libhdf5_serial.so.103 (0x00007f41867f2000)
	libstdc++.so.6 => /home/kone/dafoam/packages/miniconda3/lib/libstdc++.so.6 (0x00007f418667e000)
	libmpi_cxx.so.1 => /home/kone/dafoam/OpenFOAM/ThirdParty-v1812/platforms/linux64Gcc/openmpi-1.10.4/lib64/libmpi_cxx.so.1 (0x00007f418665d000)
	libxcb.so.1 => /lib/x86_64-linux-gnu/libxcb.so.1 (0x00007f4186633000)
	libopen-rte.so.12 => /home/kone/dafoam/OpenFOAM/ThirdParty-v1812/platforms/linux64Gcc/openmpi-1.10.4/lib64/libopen-rte.so.12 (0x00007f41865b1000)
	libnuma.so.1 => /lib/x86_64-linux-gnu/libnuma.so.1 (0x00007f41865a4000)
	libOpenCL.so.1 => /lib/x86_64-linux-gnu/libOpenCL.so.1 (0x00007f4186399000)
	libutil.so.1 => /lib/x86_64-linux-gnu/libutil.so.1 (0x00007f4186392000)
	libsz.so.2 => /lib/x86_64-linux-gnu/libsz.so.2 (0x00007f418638d000)
	libz.so.1 => /home/kone/dafoam/packages/miniconda3/lib/libz.so.1 (0x00007f418636d000)
	libXau.so.6 => /lib/x86_64-linux-gnu/libXau.so.6 (0x00007f4186367000)
	libXdmcp.so.6 => /lib/x86_64-linux-gnu/libXdmcp.so.6 (0x00007f418635f000)
	libaec.so.0 => /lib/x86_64-linux-gnu/libaec.so.0 (0x00007f4186354000)
	libbsd.so.0 => /lib/x86_64-linux-gnu/libbsd.so.0 (0x00007f418633a000)

3)  python
Python 3.7.7 (default, May  7 2020, 21:25:33) 
[GCC 7.3.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.

4) >>> import idwarp
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: /home/kone/dafoam/repos/idwarp-2.2.1/src/f2py/idwarp.so: undefined symbol: __cgns_MOD_bctypename

**Reply by friedenhe** [reply]
The problem was that your cgns lib was not linked correctly, check this line in the ldd output: "libcgns.so.3.4 => /usr/lib/x86_64-linux-gnu/libcgns.so.3.4 (0x00007f41872b3000)". It was linked to the system CGNS lib... It should link to "/home/kone/dafoam/packages/CGNS-3.3.0/lib/libcgns.so.3.3" instead. Check your CGNS installation.

**Reply by Charly-28** [reply]
Thanks  I uninstall CGNS on my compute. Now it works.
But the test is failed. Below the errors:

Argument not found. Using the default value: all
Running all tests....
--------------------------------------------------------------------------
It looks like opal_init failed for some reason; your parallel process is
likely to abort.  There are many reasons that a parallel process can
fail during opal_init; some of which are due to configuration or
environment problems.  This failure appears to be an internal failure;
here's some additional information (which may only be relevant to an
Open MPI developer):

  opal_shmem_base_select failed
  --> Returned value -1 instead of OPAL_SUCCESS

**Reply by friedenhe** [reply]
Try to run a tutorial case with one CPU core. If that works, either openmpi or mpi4py has issues.

**Reply by Charly-28** [reply]
Thanks for your reply. 
But I still have the same mistakes:

./runTests_DAPisoFoamACTL.py
--------------------------------------------------------------------------
It looks like opal_init failed for some reason; your parallel process is
likely to abort.  There are many reasons that a parallel process can
fail during opal_init; some of which are due to configuration or
environment problems.  This failure appears to be an internal failure;
here's some additional information (which may only be relevant to an
Open MPI developer):

  opal_shmem_base_select failed
  --> Returned value -1 instead of OPAL_SUCCESS

**Reply by friedenhe** [reply]
Did you load `loadDAFoam.sh` before running the job? Run these and let me know what do you see: 

`which python`

`which mpirun`

`python` and then `import mpi4py` and then `mpi4py.__path__`

**Reply by Charly-28** [reply]
1) Yes I load loadDAfoam before running job.
2) which python
./dafoam/packages/miniconda3/bin/python
3) which mpirun
./dafoam/OpenFOAM/ThirdParty-v1812/platforms/linux64Gcc/openmpi-1.10.4/bin/mpirun
4) python import mpi4py mpi4py.__path__
['/home/kone/.local/lib/python3.7/site-packages/mpi4py']

Question:
How to use mpi4py without uninstalling it locally?

Thanks

**Reply by friedenhe** [reply]
Add this line to your loadDAFoam.sh and re-load it. It will bypass the packages in .local

`export PYTHONUSERBASE=no-local-libs`

**Reply by Charly-28** [reply]
Thanks for your reply.
Now the test works but failed on the solver interfaces
Below the error:

[[55182,1],0]: A high-performance Open MPI point-to-point messaging module
was unable to find any relevant network interfaces:

Module: OpenFabrics (openib)
Host: kone-HP-Z4-G4-Workstation
Another transport will be used instead, although this may result in
lower performance.

DASimpleTFoam: Failed!

**Comment by Charly-28** [comment]
Thanks for your reply.
Now the test works but failed on the solver interfaces
Below the error:

[[55182,1],0]: A high-performance Open MPI point-to-point messaging module
was unable to find any relevant network interfaces:

Module: OpenFabrics (openib)
  Host: kone-HP-Z4-G4-Workstation

Another transport will be used instead, although this may result in
lower performance.
--------------------------------------------------------------------------
DASimpleTFoam: Failed!

**Reply by Charly-28** [reply]
Thanks for your reply.
Now the test works but failed on the solver interfaces
Below the error:

[[55182,1],0]: A high-performance Open MPI point-to-point messaging module
was unable to find any relevant network interfaces:

Module: OpenFabrics (openib)
Host: kone-HP-Z4-G4-Workstation
Another transport will be used instead, although this may result in
lower performance.

DASimpleTFoam: Failed!

**Reply by friedenhe** [reply]
Ignore this, this is a test bug on Ubuntu 20.04, but it shouldn't impact the optimization results. Try to run a tutorial.

**Reply by Charly-28** [reply]
Now all is fine.
Thank you for your availability I will contact you for the aeroacoustics.
My current project is to optimize the thrust and engine torque coefficients. 
Maybe I'll need your help. 
Thanks for your help.

Charly
